src/BRSRC13/CORE/FW/datafile.c

```diff
@@ -87,11 +295,162 @@ br_uint_32 DfStructReadBinary(br_datafile* df, br_file_struct* str, void* base)
     int c;
     int n;
     unsigned char* mp;
+    char* tmp;
     br_file_struct_member* sm;
-    struct {
+    union {
         unsigned char b[8];
         float f;
     } conv;
+
+    LOG_TRACE("(%p, %p, %p)", df, str, base);
+
+    for (m = 0; m < str->nmembers; m++) {
+
+        sm = &str->members[m];
+
+        //v5 = *((_DWORD*)sm + 1);
+        //v6 = basea;
+        //LOWORD(v7) = m
+        c = scalarTypeConvert(df, sm->type);
+        LOG_DEBUG("got c=%d", c);
+        mp = ((char*)base) + sm->offset;
+        if (c > 0x1f) {
+            continue;
+        }
+
+        switch (c) {
+        case 0:
+        case 1:
+        case 13:
+            *mp = BrFileGetChar(df->h);
+            goto LABEL_33;
+        case 2:
+        case 3:
+        case 7:
+        case 14:
+            mp[1] = BrFileGetChar(df->h);
+            *mp = BrFileGetChar(df->h);
+            LOG_DEBUG("mp=%d, mp[1]=%d", *mp, mp[1]);
+            goto LABEL_33;
+        case 4:
+        case 5:
+        case 8:
+        case 15:
+        case 24:
+        case 25:
+            mp[3] = BrFileGetChar(df->h);
+            LOG_DEBUG("got char %d", mp[3]);
+            goto LABEL_7;
+        case 6:
+            LOG_PANIC("Not implemented?");
+            conv.b[3] = BrFileGetChar(df->h);
+            conv.b[2] = BrFileGetChar(df->h);
+            conv.b[1] = BrFileGetChar(df->h);
+            conv.b[0] = BrFileGetChar(df->h);
+            *mp = (int)(conv.f * 65536.0f);
+            goto LABEL_33;
+
+        case 9:
+            mp[7] = BrFileGetChar(df->h);
+            mp[6] = BrFileGetChar(df->h);
+            mp[5] = BrFileGetChar(df->h);
+            mp[4] = BrFileGetChar(df->h);
+            mp[3] = BrFileGetChar(df->h);
+            mp[2] = BrFileGetChar(df->h);
+            mp[1] = BrFileGetChar(df->h);
+            *mp = BrFileGetChar(df->h);
+            goto LABEL_33;
+        case 10:
+        case 11:
+        case 12:
+        case 19:
+        case 20:
+        case 21:
+            LOG_WARN("Ignoring member type");
+            goto LABEL_33;
+        case 16:
+            LOG_PANIC("Not implemented?");
+            DfStructReadBinary(df, str, mp);
+            goto LABEL_33;
+        case 17:
+            for (n = 0; n < 255; n++) {
+                c = BrFileGetChar(df->h);
+                if (!c || c == -1)
+                    break;
+                tmp_string[n] = c;
+            }
+            tmp_string[n] = 0;
+            tmp = BrResStrDup(df->res ? df->res : fw.res, tmp_string);
+            // TODO: how to make sure this works if sizeof(long) != sizeof(void*)
+            *(long*)mp = tmp;
```

Comment:
Probably best to use intptr_t from <stdint.h> which will be at least sizeof(void*) provided sufficient space is allocated in what mp points to?